zebra-state/src/service.rs

```diff
@@ -1684,6 +1787,34 @@ impl Service<ReadRequest> for ReadStateService {
                 .map(|join_result| join_result.expect("panic in ReadRequest::ChainInfo"))
                 .boxed()
             }
+
+            // Used by getmininginfo, getnetworksolps, and getnetworkhashps RPCs.
+            #[cfg(feature = "getblocktemplate-rpcs")]
+            ReadRequest::CheckBlockValidity(prepared) => {
+                let timer = CodeTimer::start();
+
+                let state = self.clone();
+
+                // # Performance
+                //
+                // Allow other async tasks to make progress while concurrently reading blocks from disk.
+                let span = Span::current();
+                tokio::task::spawn_blocking(move || {
+                    span.in_scope(move || {
+                        state.check_best_chain_contextual_validity(&prepared)?;
```

Comment:
Optional: Can we have this function explicitly take an `Arc<Chain>` and a `&ZebraDb`, like all the other `read_*` functions?

It's easier to review the concurrency if we can see exactly what parts of the state the function is using.

Reply:
I re-used the same write fn but essentially did this in [`8420b75`](https://github.com/ZcashFoundation/zebra/pull/5870/commits/8420b75013c1a05ea11a9799d1f87bd338c3c27c) with your suggestion to attempt writing it to a cloned non-finalized state. 

A new block iterator would be needed to take an `Arc<Chain>` so it seems easier to rebuild a non-finalized state with only the best chain like [this](https://github.com/ZcashFoundation/zebra/pull/5630/files#diff-a2e205e7f47ebdd654a97d9a69c10fed917af8f6f6174e8cce994d5506eeded3R64-R79), but I think it should try committing it to any chain to align with the usual acceptance rules anyways.